<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Why should biologist learn programming?</title>
    <meta charset="utf-8" />
    <meta name="author" content="Guillaume Falmagne" />
    <script src="00_Why_Programming_files/header-attrs-2.28/header-attrs.js"></script>
    <link href="00_Why_Programming_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="00_Why_Programming_files/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="00_Why_Programming_files/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
    <script src="00_Why_Programming_files/clipboard-2.0.6/clipboard.min.js"></script>
    <link href="00_Why_Programming_files/xaringanExtra-clipboard-0.2.6/xaringanExtra-clipboard.css" rel="stylesheet" />
    <script src="00_Why_Programming_files/xaringanExtra-clipboard-0.2.6/xaringanExtra-clipboard.js"></script>
    <script>window.xaringanExtraClipboard(null, {"button":"Copy code <i class=\"fa fa-clipboard\"><\/i>","success":"Copied! <i class=\"fa fa-check\" style=\"color: #90BE6D\"><\/i>","error":"Not copied ðŸ˜• <i class=\"fa fa-times-circle\" style=\"color: #F94144\"><\/i>"})</script>
    <link href="00_Why_Programming_files/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
    <link href="00_Why_Programming_files/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
    <link rel="stylesheet" href="styles.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Why should biologist learn programming?
]
.subtitle[
## Biology is a computational discipline
]
.author[
### Guillaume Falmagne
]
.date[
### Sept.Â 4th, 2024
]

---




# Course structure

- Guillaume Falmagne - [g.falmagne@princeton.edu](g.falmagne@princeton.edu)
- Michelle White - [mw0425@princeton.edu](mw0425@princeton.edu)

- Website: [eeb330.github.io](http://eeb330.github.io). Contains schedule, [Google Calendar](https://calendar.google.com/calendar/u/1?cid=Y182NjA5NWU5ZTJkNjY3Y2Y2ZjQzMTdhMzU4Nzg2MTkwMDA0MDkyN2JkNWVjYjJlYTliOTgzOWUxZTI4OTFlZTM5QGdyb3VwLmNhbGVuZGFyLmdvb2dsZS5jb20), and lectures and precepts.

- Two lectures per week (Mon and Wed, 3pm)
  - Presentations and some in-class exercises

- One precept per week with Michelle (Thu 10 and 11am)

- Office hours:
  - Guyot 104a, by appointment
  - This week: special office hour, Thursday 10am in place of the precept  
    &amp;rarr; **help installing** R + python + Rstudio/VsCode + git

---

# Grading

  - **No exams**
  - Weekly-ish coding exercises (30%)
    - Graded for coherence, not correctness

  - Class and precept participation (15%)
    - Participation and in-class quizzes

  - Two coding projects (40%: 20% each)
    - Midterm and Final projects
    - Around 2 weeks to complete them

  - Open-science exercise (15%)
    - Small presentation during finals period reproducing the analysis and/or plots from a published paper. 


---
class: inverse, center, middle

# Why learn programming?

---

# Modern biology is computational biology

- Biology has become an eminently computational field &amp;ndash; as many other fields!
- Bioinformatics, statistics, handling large data sets = now basic tools of biology research and applications

![:scale 90%](figures/foamtree.png)

---

# Central part of analysis pipeline

&amp;nbsp;&amp;nbsp;

.pull-left[
What does a typical academic research project look like?
- Ideation / research question
- Literature review
- Experiments (field or lab) and **data collection**
- **Data analysis**
- **Plotting results**
- Writing paper, then submitting for publication

Similarly important in industrial projects!
]
.pull-right[
![:scale 100%](figures/pipeline_fancy.png)
]

**Need to be "fluent in coding" so that the science part stays at the center**

---
# Exponentially growing data + computing power
&lt;div class='left' style='float:left;width:43%'&gt;
![:scale 82%](figures/globaldata.webp)
![:scale 82%](figures/genomecost.jpg)
&lt;/div&gt;
&lt;div class='right' style='float:right;width:57%'&gt;
.center[
![:scale 71%](figures/Mooreslaw.png)  
]

.left[
- Easy to loose track of research goal and hypotheses
- Earlier (excellent) biologists were not trained to treat this big data...  
&amp;rarr; lack of expertise in programming and statistics at this scale  
&amp;rarr; window of opportunity for fresh mindsets
]
&lt;/div&gt;

???
Changed in last 10-20 years + researchers are constant learners

---

# Replication crisis

.pull-left[
Need clean code to share in papers, to ease:
- reproducibility of results
  - "Most published findings are false"   
  &amp;ndash; also p-hacking
- check of **correctness** (peer review)
- post publication, community can:
  - **Reuse** the code (Open Science)
  - Modify and **extend** an analysis

![](figures/reproducibility.svg)
]
.pull-right[
![:scale 110%](figures/ioannidis.png)
]

---

# Excel is very popular

![](figures/eve_online.png)

---

# Coding errors in spreadsheets reach far

.pull-left[
*Growth in the time of Debt*  
Reinhart and Rogoff, 2010, American Economic Review

- Super influential, around 5k citations
- Used to justify economic policy

![:scale 65%](figures/figure2.png)
]
.pull-right[
*Does high public debt consistently stifle economic growth? A critique of Reinhart and Rogoff*  
Thomas Herndon et al., 2014, Cambridge Journal of Economics

![](figures/excel_error.png)
]

---

# Spreadsheet are not made for data science

**Poor numerical accuracy**

Almiron et al., 2010, Journal of Statistical Software

![:scale 60%](figures/spreadsheet_tests.png)

**Black box functions**  
Poor knowledge/readability of the content of functions used in spreadsheets leads to bad science!

---
class: inverse, center, middle

# What and how to learn programming?

---

# Bespoke Custom Analysis

.pull-left[
- Custom program for your custom analysis:
  - is more likely to solve *your* exact question
  - does what you designed it for 
- Canned solutions might solve some other problem only *related* to what you really care about.

- Bespoke software solution can then be adapted to other situations via code sharing or generic packages:
  - For you
  - From you
]
.pull-right[
![](figures/Tailor.png)
]
---

# Why R and/or python?


.pull-left[

- Great package ecosystem  
(CRAN, conda, bioconductor, ...)
- Easy to learn
- Good balance of easy-to-write (and to read)  
and fast-to-run code
- Both can easily call code from faster languages  
(e.g. C++) if needed
- Widely used 
]
.pull-right[
  ![:scale 80%](figures/pythonxkcd.png)
]
---

# Effective Visualization 

- Visualization is pivotal for understanding and communicating data and results.

.pull-left[
- R offers great visualization libraries, such as ggplot2.


``` r
library(ggplot2)
library(cowplot)
ggplot(iris, aes(Sepal.Width, Sepal.Length, color = Species)) +
geom_point() + stat_ellipse() + theme_cowplot()
```

&lt;img src="00_Why_Programming_files/figure-html/iris plot-1.png" height="300" /&gt;
]
.pull-right[
- In Python: matplotlib

![](figures/Matplotlib.png)
]

---

# Take home

.pull-left[
- Reproducible research ensures **transparency and verifiability**.

- Coding functions and your model yourself (in R, Python, ...) allows for **tailored analysis + independence and creativity** in methods.

- Effective visualization enhances **data understanding and communication**.

- Sharing analytical tools and leveraging package ecosystems promotes **collaboration (Open Science)**.
]
.pull-right[

### Suggested Readings

- [**A (very) short introduction to R**](https://cran.r-project.org/doc/contrib/Torfs+Brauer-Short-R-Intro.pdf)
- [**Python Introduction, W3schools**](https://www.w3schools.com/python/python_intro.asp)
- [Python cheat sheet](https://perso.limsi.fr/pointal/_media/python:cours:mementopython3-english.pdf)
- [British Ecological Society Guides](https://www.britishecologicalsociety.org/publications/guides-to/)
- [Regular Expressions in R](https://r4ds.hadley.nz/regexps)
- [Data Wrangling Cheat Sheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf)

Longer texts

- [R for Data Science](https://r4ds.hadley.nz/)
- [Bioinformatics Data Skills](https://vincebuffalo.com/book/)
- [Advanced R](https://adv-r.hadley.nz/)
- [Mastering Regular Expressions](http://regex.info/book.html)
- [ggplot2 book](https://ggplot2-book.org/)
]

---
class: inverse, center, middle

# What to expect in this course?

---

# Content to expect 

.pull-left[
**What we are covering**
- Basic interactions with a programming language
- Version control
- Some computer science concepts, like control flow and types
- Data science skills, including basic statistics
- Data visualization
- Bioinformatics tools and genomic pipelines + sequence analysis
- Using clusters
]

--

.pull-right[
**What we are *not* covering**

- Advanced statistics and data analysis
- Logics and algorithms
- CS theory
]

---

# Objectives of the course: teach you to teach yourself

.pull-left[
- Programming ***has to be*** learned by doing, googling, leaving the comfort zone

![:scale 90%](figures/stackoverflow_lost.jpg)

]
.pull-right[

- Goal of the course: **primer in multiple essential tools**, so that you:
  - can code simple programs independently 
  - get the confidence and skills to **navigate the jungle of available resources**
  - can **learn yourself** the code to adress your specific questions in biological science
  - do this faster thanks to shared experience

![:scale 60%](figures/owl.png)
]

---
# Prerequisites: to be installed

Check the website [eeb330.github.io](http://eeb330.github.io) for instructions to install:
- R
- Python
- VsCode (preferred) *OR* Rstudio
- Git and github account

Help proposed Thursday 10am (Guyot 104a) if problems
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="macros.js"></script>
<script>var slideshow = remark.create({
"highlightLines": true,
"chakra": "libs/remark-latest.min.js",
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
(function(time) {
  var d2 = function(number) {
    return ('0' + number).slice(-2); // left-pad 0 to minutes/seconds
  },

  time_format = function(total) {
    var secs = Math.abs(total) / 1000;
    var h = Math.floor(secs / 3600);
    var m = Math.floor(secs % 3600 / 60);
    var s = Math.round(secs % 60);
    var res = d2(m) + ':' + d2(s);
    if (h > 0) res = h + ':' + res;
    return res;  // [hh:]mm:ss
  },

  slide_number_div = function(i) {
    return document.getElementsByClassName('remark-slide-number').item(i);
  },

  current_page_number = function(i) {
    return slide_number_div(i).firstChild.textContent;  // text "i / N"
  };

  var timer = document.createElement('span'); timer.id = 'slide-time-left';
  var time_left = time, k = slideshow.getCurrentSlideIndex(),
      last_page_number = current_page_number(k);

  setInterval(function() {
    time_left = time_left - 1000;
    timer.innerHTML = ' ' + time_format(time_left);
    if (time_left < 0) timer.style.color = 'red';
  }, 1000);

  slide_number_div(k).appendChild(timer);

  slideshow.on('showSlide', function(slide) {
    var i = slide.getSlideIndex(), n = current_page_number(i);
    // reset timer when a new slide is shown and the page number is changed
    if (last_page_number !== n) {
      time_left = time; last_page_number = n;
      timer.innerHTML = ' ' + time_format(time); timer.style.color = null;
    }
    slide_number_div(i).appendChild(timer);
  });
})(60000);
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
