<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Data I/O</title>
    <meta charset="utf-8" />
    <meta name="author" content="Diogo Melo" />
    <script src="06_DataIO_files/header-attrs-2.21/header-attrs.js"></script>
    <link href="06_DataIO_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="06_DataIO_files/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="06_DataIO_files/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="styles.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Data I/O
]
.subtitle[
## Reading and writting biological data
]
.author[
### Diogo Melo
]
.date[
### Sept. 25th, 2023
]

---


# To analyze the data, you must first be the data

- Almost all projects begin with data going into software

- All of them end with output being extracted

- We have to learn how to read and write from files

---

# Basic data types 

- Tabular

    - Most common type in ecology.
    - Usually goes into a data.frame.
    - Excel files, csv, tsv, fixed width...

- Genomic data

    - Most common in genetics, bioinformatics
    - Usually pretty large, so requires some care and specialized data structures
    - Genotype data, sequence data, aligned sequence data, expression data

- Databases 

    - Relational data that is more complex than a single table

---

# Tabular data


```r
head(iris)
```

```
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species Petal.Length.Class
## 1          5.1         3.5          1.4         0.2  setosa              Short
## 2          4.9         3.0          1.4         0.2  setosa              Short
## 3          4.7         3.2          1.3         0.2  setosa              Short
## 4          4.6         3.1          1.5         0.2  setosa              Short
## 5          5.0         3.6          1.4         0.2  setosa              Short
## 6          5.4         3.9          1.7         0.4  setosa              Short
```


---


# Reading tabular data

.pull-left[
- Base functions:
    - Output data.frames
    - Names have `.` in them, e.g. read.table, read.csv

- `readr` package
    - Mostly drop-in replacements for the base functions, but with some more (or less) sensible choices
    - ties into the `tidyverse` groups of packages we will see later on
        - there are super popular, so important to know.
    - Names have `_` in them, e.g. read_delim, read_csv

]
.pull-right[

- `data.table` package
    - Notable because it's `fread` function is so blazing fast it is sometimes the only choice for large data

- `rio` package
    - sort of jack of all traits magic package that attempts to figure out the correct way to read your data from the file extension
    - Only has one generic `import` function that mostly get's it right.
]

---



# read.table

.pull-left[
- General purpose table reading function.
- Reads a file in table format and creates a data.frame.


```r
# Basic Usage
my_data &lt;- read.table("my_file.tsv", 
                      header = TRUE, 
                      sep = "\t")
head(my_data)
```

```
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
## 3          4.7         3.2          1.3         0.2  setosa
## 4          4.6         3.1          1.5         0.2  setosa
## 5          5.0         3.6          1.4         0.2  setosa
## 6          5.4         3.9          1.7         0.4  setosa
```
]
---

# read.csv &amp; read.csv2

Special cases of read.table designed for CSV files.

- read.csv: assumes field separator is "," and . for decimal point.
- read.csv2: assumes field separator is ";" and , for decimal point (common in European datasets).

```r
data_csv &lt;- read.csv("my_data.csv")

# For European-style CSV
data_csv2 &lt;- read.csv2("my_data_european.csv")
```
![](figures/csv_ex.png)

---

# readr package

- Outputs a tibble, a kind of modern data.frame with some quirks
    - does some smart figuring-out of column types
    - avoids most conversions to factor


```r
# Basic Usage
library(readr)
my_data_tbl &lt;- read_delim("my_file.tsv", 
                           delim = "\t")
```

```
## Rows: 150 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
## Delimiter: "\t"
## chr (1): Species
## dbl (4): Sepal.Length, Sepal.Width, Petal.Length, Petal.Width
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
```

---

# tibble data.frame


```r
my_data_tbl
```

```
## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # ℹ 140 more rows
```

---

# data.table's fread

- data.table is a very fast package to deal with tabular data
- Has its own very particular syntax
- Can deal with datasets above 100GB easily


```r
library(data.table)
data_dt &lt;- fread("my_file.csv")
data_dt
```

```
##      Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
##   1:          5.1         3.5          1.4         0.2    setosa
##   2:          4.9         3.0          1.4         0.2    setosa
##   3:          4.7         3.2          1.3         0.2    setosa
##   4:          4.6         3.1          1.5         0.2    setosa
##   5:          5.0         3.6          1.4         0.2    setosa
##  ---                                                            
## 146:          6.7         3.0          5.2         2.3 virginica
## 147:          6.3         2.5          5.0         1.9 virginica
## 148:          6.5         3.0          5.2         2.0 virginica
## 149:          6.2         3.4          5.4         2.3 virginica
## 150:          5.9         3.0          5.1         1.8 virginica
```
---

# Reading data of the interwebs


```r
input &lt;- if (file.exists("flights14.csv")) {
   "flights14.csv"
} else {
  "https://raw.githubusercontent.com/Rdatatable/data.table/master/vignettes/flights14.csv"
}
flights &lt;- fread(input)
flights
```

```
##         year month day dep_delay arr_delay carrier origin dest air_time
##      1: 2014     1   1        14        13      AA    JFK  LAX      359
##      2: 2014     1   1        -3        13      AA    JFK  LAX      363
##      3: 2014     1   1         2         9      AA    JFK  LAX      351
##      4: 2014     1   1        -8       -26      AA    LGA  PBI      157
##      5: 2014     1   1         2         1      AA    JFK  LAX      350
##     ---                                                                
## 253312: 2014    10  31         1       -30      UA    LGA  IAH      201
## 253313: 2014    10  31        -5       -14      UA    EWR  IAH      189
## 253314: 2014    10  31        -8        16      MQ    LGA  RDU       83
## 253315: 2014    10  31        -4        15      MQ    LGA  DTW       75
## 253316: 2014    10  31        -5         1      MQ    LGA  SDF      110
##         distance hour
##      1:     2475    9
##      2:     2475   11
##      3:     2475   19
##      4:     1035    7
##      5:     2475   13
##     ---              
## 253312:     1416   14
## 253313:     1400    8
## 253314:      431   11
## 253315:      502   11
## 253316:      659    8
```
---


# Genomics data

- FASTA - Simple collections of named DNA/protein sequences

- FASTQ - Extension of FASTA format, contains additional quality information. Widely used for storing unaligned sequencing reads

- SAM/BAM - Alignments of sequencing reads to a reference genome

- BED - Region-based genome annotation information (e.g. a list of
genes and their genomic locations). Used for visualization or simple
enumeration

- GFF/GTF - gene-centric annotations

- (big)WIG - continuous signal representation

- VCF - variant call format, to store information about genomic variants

---

# Reference genome as a set of coordinates

Most genomics data formats have some sort of coordinate information

.center[
    ![](figures/coordinates.png)
]

---


# DNA Sequence data

.pull-left[
### FASTA

- A sequence in FASTA format consists of:
 - One line starting with a "&gt;" sign, followed by a sequence identification code.

 - One or more lines containing the sequence itself.
```
        A --&gt; adenosine           M --&gt; A C (amino)
        C --&gt; cytidine            S --&gt; G C (strong)
        G --&gt; guanine             W --&gt; A T (weak)
        T --&gt; thymidine           B --&gt; G T C
        U --&gt; uridine             D --&gt; G A T
        R --&gt; G A (purine)        H --&gt; A C T
        Y --&gt; T C (pyrimidine)    V --&gt; G C A
        K --&gt; G T (keto)          N --&gt; A G C T (any)
                                  -  gap of indeterminate length
```
]
.pull-right[
### Human genome

[GRCh38 Genome Reference Consortium Human Reference 38](https://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/)
![](figures/human_genome.png)
]

---

# Next Generaton Sequencing scheme

.center[
![:scale 40%](figures/NGS-rainbow-sequence.webp)
]

.ref[https://info.abmgood.com/next-generation-sequencing-ngs-data-analysis]

---

# Next Generaton Sequencing scheme

![:scale 85%](figures/NGS.jpg)

.ref[https://irepertoire.com/ngs-overview-from-sample-to-sequencer-to-results/]

---


# Short read data - FASTQ

- An intermediate filetype that is used for downstream analysis
- Usually we just pipe it to allignment software and never look at it

.ref[https://knowledge.illumina.com/software/general/software-general-reference_material-list/000002211]

### General structure

- `@` line with an alphanumeric ID
- Base pair sequence
- `+` 
- Quality score for each read position

```
	@SIM:1:FCX:1:15:6329:1045 1:N:0:2
	TCGCACTCAACGCCCTGCATATGACAAGACAGAATC
	+
	&lt;&gt;;##=&gt;&lt;9=AAAAAAAAAA9#:&lt;#&lt;;&lt;&lt;&lt;????#=
```

This is the quality score rank:

```
!"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~
```

---

# Single or pair-end reads

.center[
![:scale 50%](figures/paired_end.png)
]

---

# Read depth and coverage

.center[
![](figures/depth.jpeg)
]

.ref[https://journals.sagepub.com/doi/10.1177/1099800417750746]
---

# Shotgun assemmbly

.center[
![](figures/assembly.webp)
]
.ref[https://www.differencebetween.com/difference-between-clone-by-clone-sequencing-and-shotgun-sequencing/]

# Alignments - SAM and BAM

---

# Some sequencing methods

.center[
![:scale 50%](figures/seq_comparison.jpg)
]
.ref[Lou, R. N., Jacobs, A., Wilder, A. P. &amp; Therkildsen, N. O. A beginner’s guide to low-coverage whole genome sequencing for population genomics. Mol. Ecol. 30, 5966–5993 (2021) [10.1111/mec.16077](https://onlinelibrary-wiley-com.ezproxy.princeton.edu/share/YHRVMJNPQCIA9BPMKSJU?target=10.1111/mec.16077)]

---

# Genotyping data

[The Variant Call Format (VCF)](https://samtools.github.io/hts-specs/VCFv4.2.pdf)

![:scale 100%](figures/vcf_example.png)
 
---

# Common VCF formats

```
##INFO=&lt;ID=AF,Number=A,Type=Float,Description="Estimated Allele Frequencies"&gt;
##INFO=&lt;ID=AR2,Number=1,Type=Float,Description="Allelic R-Squared: estimated correlation between most probable ALT dose and true ALT dose"&gt;
##INFO=&lt;ID=DR2,Number=1,Type=Float,Description="Dosage R-Squared: estimated correlation between estimated ALT dose [P(RA) + 2*P(AA)] and true ALT dose"&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;
##FORMAT=&lt;ID=DS,Number=1,Type=Float,Description="estimated ALT dose [P(RA) + P(AA)]"&gt;
##FORMAT=&lt;ID=GP,Number=G,Type=Float,Description="Estimated Genotype Probability"&gt;
```

```
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	127 ...
1	5385887	AX-168417668	C	G	.	PASS	AR2=1;DR2=1;AF=0.165	GT:DS:GP	0|0:0:1,0,0	...
```

---

# Reading VCFs in R

.pull-left[

```r
library(vcfR)
vcf = read.vcfR("sample.vcf")
```

```
## Scanning file to determine attributes.
## File attributes:
##   meta lines: 21
##   header_line: 22
##   variant count: 9
##   column count: 12
## 
Meta line 21 read in.
## All meta lines processed.
## gt matrix initialized.
## Character matrix gt created.
##   Character matrix gt rows: 9
##   Character matrix gt cols: 12
##   skip: 0
##   nrows: 9
##   row_num: 0
## 
Processed variant: 9
## All variants processed
```
]

---

# Reading VCFs in R

```r
str(vcf)
```

```
## Formal class 'vcfR' [package "vcfR"] with 3 slots
##   ..@ meta: chr [1:21] "##fileformat=VCFv4.0" "##fileDate=20090805" "##source=myImputationProgramV3.1" "##reference=1000GenomesPilot-NCBI36" ...
##   ..@ fix : chr [1:9, 1:8] "19" "19" "20" "20" ...
##   .. ..- attr(*, "dimnames")=List of 2
##   .. .. ..$ : NULL
##   .. .. ..$ : chr [1:8] "CHROM" "POS" "ID" "REF" ...
##   ..@ gt  : chr [1:9, 1:4] "GT:HQ" "GT:HQ" "GT:GQ:DP:HQ" "GT:GQ:DP:HQ" ...
##   .. ..- attr(*, "dimnames")=List of 2
##   .. .. ..$ : NULL
##   .. .. ..$ : chr [1:4] "FORMAT" "NA00001" "NA00002" "NA00003"
```

---

# Using fread


```r
df_vcf = fread(file='sample.vcf', sep='\t', header = TRUE, skip = '#CHROM')
df_vcf
```

```
##    #CHROM     POS        ID REF    ALT QUAL FILTER
## 1:     19     111         .   A      C  9.6      .
## 2:     19     112         .   A      G   10      .
## 3:     20   14370 rs6054257   G      A   29   PASS
## 4:     20   17330         .   T      A    3    q10
## 5:     20 1110696 rs6040355   A    G,T   67   PASS
## 6:     20 1230237         .   T      .   47   PASS
## 7:     20 1234567 microsat1   G GA,GAC   50   PASS
## 8:     20 1235237         .   T      .    .      .
## 9:      X      10    rsTest  AC  A,ATG   10   PASS
##                                 INFO      FORMAT        NA00001        NA00002
## 1:                                 .       GT:HQ      0|0:10,10      0|0:10,10
## 2:                                 .       GT:HQ      0|0:10,10      0|0:10,10
## 3:           NS=3;DP=14;AF=0.5;DB;H2 GT:GQ:DP:HQ 0|0:48:1:51,51 1|0:48:8:51,51
## 4:               NS=3;DP=11;AF=0.017 GT:GQ:DP:HQ 0|0:49:3:58,50   0|1:3:5:65,3
## 5: NS=2;DP=10;AF=0.333,0.667;AA=T;DB GT:GQ:DP:HQ 1|2:21:6:23,27   2|1:2:0:18,2
## 6:                   NS=3;DP=13;AA=T GT:GQ:DP:HQ 0|0:54:.:56,60 0|0:48:4:51,51
## 7:        NS=3;DP=9;AA=G;AN=6;AC=3,1    GT:GQ:DP        0/1:.:4       0/2:17:2
## 8:                                 .          GT            0/0            0|0
## 9:                                 .          GT              0            0/1
##         NA00003
## 1:      0/1:3,3
## 2:      0/1:3,3
## 3: 1/1:43:5:.,.
## 4: 0/0:41:3:.,.
## 5: 2/2:35:4:.,.
## 6: 0/0:61:2:.,.
## 7:     1/1:40:3
## 8:          ./.
## 9:          0|2
```

---

# Gene expression

.center[
![:scale 40%](figures/central_dogma_sketch.webp)
]
---

# Gene expression

.center[
![](figures/central_sketch.png)
]
---

# Gene expression

![](figures/Central-dogma.jpg)

---

# Gene expression data (RNAseq)

.pull-left[
![:scale 90%](figures/rna_seq.jpg)
.ref[https://www.nature.com/articles/nrg2484]
]

.pull-right[
- Alligned RNA reads that map to genes are contributed

- The counts are proportional to RNA expression


```r
counts &lt;- read.delim("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2018-June-RNA-Seq-Workshop/master/thursday/all_counts.txt")
counts[1:10, 1:7]
```

```
##                  C61 C62 C63 C64  C91  C92 C93
## AT1G01010        289 317 225 343  325  449 310
## AT1G01020        127  78 142 130  156  146 144
## AT1G03987          0   0   0   0    0    0   0
## AT1G01030         17  25  32  24   22   25  21
## AT1G01040        605 415 506 565  762  854 658
## AT1G03993          1   1   0   0    0    0   1
## AT1G01046          0   0   0   0    0    0   0
## ENSRNA049757489    0   0   0   0    0    0   0
## AT1G01050       1164 876 935 979 1146 1254 948
## AT1G03997          0   0   0   0    0    0   0
```

]

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="macros.js"></script>
<script>var slideshow = remark.create({
"highlightLines": true,
"chakra": "libs/remark-latest.min.js",
"ratio": "16:10"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
(function(time) {
  var d2 = function(number) {
    return ('0' + number).slice(-2); // left-pad 0 to minutes/seconds
  },

  time_format = function(total) {
    var secs = Math.abs(total) / 1000;
    var h = Math.floor(secs / 3600);
    var m = Math.floor(secs % 3600 / 60);
    var s = Math.round(secs % 60);
    var res = d2(m) + ':' + d2(s);
    if (h > 0) res = h + ':' + res;
    return res;  // [hh:]mm:ss
  },

  slide_number_div = function(i) {
    return document.getElementsByClassName('remark-slide-number').item(i);
  },

  current_page_number = function(i) {
    return slide_number_div(i).firstChild.textContent;  // text "i / N"
  };

  var timer = document.createElement('span'); timer.id = 'slide-time-left';
  var time_left = time, k = slideshow.getCurrentSlideIndex(),
      last_page_number = current_page_number(k);

  setInterval(function() {
    time_left = time_left - 1000;
    timer.innerHTML = ' ' + time_format(time_left);
    if (time_left < 0) timer.style.color = 'red';
  }, 1000);

  slide_number_div(k).appendChild(timer);

  slideshow.on('showSlide', function(slide) {
    var i = slide.getSlideIndex(), n = current_page_number(i);
    // reset timer when a new slide is shown and the page number is changed
    if (last_page_number !== n) {
      time_left = time; last_page_number = n;
      timer.innerHTML = ' ' + time_format(time); timer.style.color = null;
    }
    slide_number_div(i).appendChild(timer);
  });
})(60000);
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
